### PR TITLE
Add support for sts assume_role tokens

### PR DESCRIPTION
**Pull request by UnrealAkama**
AWS allows users to authenticate requests with temporary credentials that are requested from the STS service. Doing so requires that the token header is signed and included with the request. This adds an optional parameter to calls to allow for a token parameter to be passed in.

This is working for me and I've successfully incorporated into my code but I'm putting this in draft until I can test it a bit more. Please let me know if you would like anything changed. 

I believe this might solve #76 as well.

### CODE REVIEW

File: async/runtime.mli [hunk 1]
```diff
@@ -35,6 +35,7 @@ val run_request :
      region:string
   -> access_key:string
   -> secret_key:string
+  -> ?token:string
```
Comment (tmcgilchrist) [inline async/runtime.mli:38]: It seems that a request can have either `token` or a `access_key/secret_key`, the type on this function should indicate this. Something like:

``` ocaml
type security_token = 
 | AccessKey string string
 | Token string
```
would be an improvement.

Could you also update the comment associated with this function to mention it now supports a token?
I've got document generation available at http://inhabitedtype.github.io/ocaml-aws/

File: lib/aws.ml [hunk 2]
```diff
@@ -501,7 +501,7 @@ module Signing = struct
   (* NOTE(dbp 2015-01-13): This is a direct translation of reference implementation at:
    * http://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html
    *)
-  let sign_request ~access_key ~secret_key ~service ~region (meth, uri, headers) =
+  let sign_request ~access_key ~secret_key ?token ~service ~region (meth, uri, headers) =
```
Comment (tmcgilchrist) [inline lib/aws.ml:504]: Same suggestion here with either `token` or a `access_key/secret_key` being required.

File: libraries/s3/lib_test/test_async.ml [hunk 3]
```diff
@@ -3,11 +3,13 @@ open Aws_s3_test
 module T = TestSuite (struct
   type 'a m = 'a Async.Deferred.t
 
-  let access_key = Unix.getenv "AWS_ACCESS_KEY"
+  let access_key = Unix.getenv "AWS_ACCESS_KEY_ID"
 
-  let secret_key = Unix.getenv "AWS_SECRET_KEY"
+  let secret_key = Unix.getenv "AWS_SECRET_ACCESS_KEY"
```
Comment (tmcgilchrist) [inline libraries/s3/lib_test/test_async.ml:8]: Happy to change these values but could it be done in another PR and the change made across the board for all `lib_test` tests.

**Comment by tmcgilchrist** [comment]
Finally could you add a test that does something basic like creating a token and then using it on a request.
Right now I run everything in `lib_test` manually against my AWS credentials and check that it passes. 

Thanks for the code :-)

File: async/runtime.mli (hunk 1 above)
Comment (akringwood) [reply async/runtime.mli:38]: So the token works by providing an additional token on top of a key/secret. The flow being that you call sts assume-role or some other endpoint, get a new key, secret, and token. You can then use all three of those to authenticate requests until the credentials expire.

There is a bit more information here https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_use-resources.html which covers some of the main points but a bunch of the information is spread across a wide range of their docs.

**Comment by akringwood** [comment]
So I don't quite know the exact best way to conduct the tests for this because the IAM api isn't generated yet which makes it hard to create/delete all the resources we would need to do a clean test. I suspect the best way might be to add some code that allows for the token to be passed in optionally on via env to the tests and support running the tests with that token. Then you would just need a role in your aws account that you test with and you could assume-role and run the tests.

File: async/runtime.mli (hunk 1 above)
Comment (tmcgilchrist) [reply async/runtime.mli:38]: Got it, thanks for the explanation. 
Happy with the code as it stands in that case.

**Comment by tmcgilchrist** [comment]
As a suggestion can you create the resources needed via the AWS cli and then pass those through to the OCaml code to perform the request. Longer term we would have IAM bindings and replace the AWS cli part.

> I suspect the best way might be to add some code that allows for the token to be passed in optionally on via env to the tests and support running the tests with that token. Then you would just need a role in your aws account that you test with and you could assume-role and run the tests.

I think that matches my suggestion above?

**Comment by tmcgilchrist** [comment]
@UnrealAkama Are you able to finish off this change? 
There should be working STS support in OCaml now, I fixed a few serialisation issues and the `module AssumeRole` seems to be working for the simple tests I've done with it but as you said SSM isn't there yet. Maybe having a manual setup via the AWS cli and then running a single test, until I have time to generate SSM bindings.

**Comment by UnrealAkama** [comment]
Yah, that makes sense to me. I'll get a commit up in the next couple days that has a way to test it.

**Comment by UnrealAkama** [comment]
Added in #117.
